src/websocket/server.cc

```diff
@@ -183,23 +267,65 @@ future<> connection::read_loop() {
         });
     }).then_wrapped([this] (future<> f) {
         if (f.failed()) {
-            wlogger.error("Read failed: {}", f.get_exception());
+            wlogger.error("Failure: {}", f.get_exception());
         }
         return _replies.push_eventually({});
     }).finally([this] {
         return _read_buf.close();
     });
 }
 
+future<> connection::send_data(temporary_buffer<char> buff) {
+    sstring data;
+    data.append("\x81", 1);
+    if ((126 <= buff.size()) && (buff.size() <= std::numeric_limits<uint16_t>::max())) {
+        uint16_t length = buff.size();
+        length = htobe16(length);
+        data.append("\x7e", 1);
+        data.append(reinterpret_cast<char*>(&length), sizeof(uint16_t));
+    } else if (std::numeric_limits<uint16_t>::max() < buff.size()) {
+        uint64_t length = buff.size();
+        length = htobe64(length);
+        data.append("\x7f", 1);
+        data.append(reinterpret_cast<char*>(&length), sizeof(uint64_t));
+    } else {
+        uint8_t length = buff.size() & 0x7F;
+        data.append(reinterpret_cast<char*>(&length), sizeof(uint8_t));
+    }
+    data.append(buff.get(), buff.size()); 
+    temporary_buffer<char> result{data.data(), data.length()}; 
```

Comment:
All of this copying (all the appends and now this final copy when creating the temporary buffer) is fairly inefficient. I wonder if we can't more efficiently create a temporary_buffer with the approximate size (buff.size() plus I think 9 (?) bytes?) and then potentially trim() a few bytes from the end if we need a few bytes less.

Reply:
Right, makes sense. What would also make sense is to simply avoid sending a single contiguous buffer and instead either just call `write()` two times - once for the header and once with the `buff` that we already have, or even as a single `write()`, but with two buffers wrapped in a [scattered_message](https://github.com/scylladb/seastar/blob/master/include/seastar/core/scattered_message.hh).